File: java/shared/java/org/signal/libsignal/zkgroup/GenericServerSecretParams.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup;

import java.security.SecureRandom;
import org.signal.libsignal.zkgroup.internal.ByteArray;
import org.signal.libsignal.internal.Native;

import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

public final class GenericServerSecretParams extends ByteArray {

  public static GenericServerSecretParams generate() {
    return generate(new SecureRandom());
  }

  public static GenericServerSecretParams generate(SecureRandom secureRandom) {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.GenericServerSecretParams_GenerateDeterministic(random);

    try {
      return new GenericServerSecretParams(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    } 
  }

  public GenericServerSecretParams(byte[] contents) throws InvalidInputException  {
    super(contents);
    Native.GenericServerSecretParams_CheckValidContents(contents);
  }

  public GenericServerPublicParams getPublicParams() {
    byte[] newContents = Native.GenericServerSecretParams_GetPublicParams(contents);
    try {
      return new GenericServerPublicParams(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    } 
  }

}

File: node/ts/zkgroup/GenericServerSecretParams.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { randomBytes } from 'node:crypto';
import * as Native from '../Native.js';
import ByteArray from './internal/ByteArray.js';

import { RANDOM_LENGTH } from './internal/Constants.js';
import GenericServerPublicParams from './GenericServerPublicParams.js';

export default class GenericServerSecretParams extends ByteArray {
  private readonly __type?: never;

  static generate(): GenericServerSecretParams {
    const random = randomBytes(RANDOM_LENGTH);

    return GenericServerSecretParams.generateWithRandom(random);
  }

  static generateWithRandom(random: Uint8Array): GenericServerSecretParams {
    return new GenericServerSecretParams(
      Native.GenericServerSecretParams_GenerateDeterministic(random)
    );
  }

  constructor(contents: Uint8Array) {
    super(contents, Native.GenericServerSecretParams_CheckValidContents);
  }

  getPublicParams(): GenericServerPublicParams {
    return new GenericServerPublicParams(
      Native.GenericServerSecretParams_GetPublicParams(this.contents)
    );
  }
}
